### DISCUSSION TITLE
[path repository] Composer keeps removing package and reinstalls it from upstream

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#11509] [Q&A / Support]
We have a private Composer packagist that adds explicit version information to every package that could be the root cause of this behavior...

What we are currently experimenting with how we can use Composer for installing packages from an upstream repo (e,g. foo/bar:1.0.0), but after they are installed they would also become available via a local path repository for Composer. The goal is allowing package content customization on downstream (without patching, etc.) without completely breaking connection with upstream. So after foo/bar:1.0.0 was installed, the locally available foo/bar:1.0.0 may not be the same as in upstream - and we would be fine with that behavior since every changes would be tracked in VCS.

While I was testing this behavior I spotted that if a package has `version` information in composer.json and available in two locations (upstream and a local path repo)  `composer update --lock` keeps removing and re-installing the package. Why that is happening? Some kind of hash checking is failing behind the scenes?

I have managed to put together a small POC with only publicly available packages: https://github.com/mxr576/composer-issue-path-repo1

```shell
$ composer --version
Composer version 2.5.8 2023-06-09 17:13:21

$ git rev-parse HEAD
963343ea1d4a864fc2ab4249a041c235fc616e93
$ composer update --lock
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Generating autoload files

## Add version info to the local package - could have been defined by upstream as well.
$ git rev-parse HEAD
b80ef836c78b75c747387cf0412da3c645b1225c
$ composer update --lock
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 0 updates, 0 removals
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Removing su-sws/stanford_basic (8.5.19)
  - Installing su-sws/stanford_basic (8.5.19): Source already present
$ ls web/themes/custom/
# THE PACKAGE IS MISSING.

# Re-running the command installs the package from upstream.
$ composer update --lock
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 0 updates, 0 removals
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing su-sws/stanford_basic (8.5.19): Extracting archive
$ ls web/themes/custom/
stanford_basic
```

**Comment by mxr576** [comment]
So the `composer update --lock` that performs a removal and an installation just tries to ensure that composer.lock refers to the actual source of the package... and the party is killed by a sanity check in `\Composer\Downloader\PathDownloader::install()` which of course does not want to re-install a package that is already there, but is it, since there was a remove operation...

```diff
         {
             "name": "su-sws/stanford_basic",
             "version": "8.5.19",
-            "source": {
-                "type": "git",
-                "url": "https://github.com/SU-SWS/stanford_basic.git",
-                "reference": "deb14a4d72f862bfc295e08ca9bf8407b4f615c5"
-            },
             "dist": {
-                "type": "zip",
-                "url": "https://api.github.com/repos/SU-SWS/stanford_basic/zipball/deb14a4d72f862bfc295e08ca9bf8407b4f615c5",
-                "reference": "deb14a4d72f862bfc295e08ca9bf8407b4f615c5",
-                "shasum": ""
+                "type": "path",
+                "url": "web/themes/custom/stanford_basic",
+                "reference": "e267981cbd7686fc73f1ea91e6f68f12e5ebb3ff"
             },

```

**Comment by mxr576** [comment]
Okay, so it looks like this is fundamentally a chicken or egg like problem... we are trying to tell Composer to _replace_ an already installed package that was installed on location X, with a local package that is currently installed on location X. 

For the first sight  `\Composer\Installer\LibraryInstaller::updateCode()` does what it should in a regular scenario: when package provider (downloader) changes it removes the already installed package to ensure that there will be no remaining remnants; then it tries to install the package with the new provider (downloader). 

```
        // if downloader type changed, or update failed and user asks for reinstall,
        // we wipe the dir and do a new install instead of updating it
        $promise = $initialDownloader->remove($initial, $targetDir);
```

**Comment by stof** [comment]
Having a `path` repository being sourced from the place where composer installs packages is indeed not a supported case.

**Reply by mxr576** [reply]
Do you have an ad hoc idea how the described use case can be solved with "built-in" features?

What I am trying to avoid is building something custom that could be partially like the unpack feature of Symfony Flex (this is what we would be actually needed) with an extra, that keeps the unpacked dependency somewhere in the project so Drupal can grab some code from those too.
